lightning-invoice/src/payment.rs

```diff
@@ -733,8 +733,69 @@ where
 			_ => {},
 		}
 
-		// Delegate to the decorated event handler unless the payment is retried.
-		self.event_handler.handle_event(event)
+		Some(event)
+	}
+}
+
+/// A utility for paying [`Invoice`]s and sending spontaneous payments that intercepts relevant
+/// events and forwards them to the EventHandler.
+///
+/// See [module-level documentation] for details.
+///
+/// [module-level documentation]: crate::payment
+pub struct InvoicePayerEventHandler<P: Deref, R: Router, L: Deref, T: Time, E: EventHandler>
+where
+	P::Target: Payer,
+	L::Target: Logger,
+{
+	/// A utility for paying [`Invoice`]s and sending spontaneous payments.
+	pub invoice_payer: InvoicePayerUsingTime<P, R, L, T>,
+	/// An event handler to continue handling any events intercepted by the InvoicePayer.
+	pub event_handler: E,
+}
```

Comment:
Hmm... this does't make for great UX when calling methods on `InvoicePayer`. Isn't clear to me why this is required. Could you clarify?

Reply:
I guess it doesn't need to be tuple. Just seems like exposing the wrapper is unnecessary if it holds a reference instead.

Reply:
To clarify, I'm saying just hide the wrapper in `BackgroundProcessor` if `InvoicePayer` no longer implements `EventHandler`. Each the user and `BackgroundProcessor` would have a reference to `InvoicePayer`. The user for paying and the `BackgroundProcessor` for delegating event handling.

Reply:
We could implement `Deref` on it to avoid needing to access the `invoice_payer` field. I'd rather have methods on `InvoicePayer` consume `self`, pass an `{Async}EventHandler`, and return the `Deref`able wrapper than exposing the details to the user.

Another, perhaps preferable idea: Keep the `E` type parameter on `InvoicePayer` and have two different `impl` blocks for `new` each with different trait bounds.... probably would need different function names. And similarly, only implement `EventHandler` and the async trait on `impl` blocks with the respective trait bounds. I *think* this would work though each trait may need a common supertrait like we do in `Features` (not sure).

Reply:
> Not sure I understood this?

Don't have the user manually construct a wrapper using `pub` fields. Have a method on `InvoicePayer` do it instead.
> 
> Hmm, ugh. Let's try to keep the class hierarchy simply for now, and clean it all up by moving retry logic into ChannelManager instead.

This doesn't change the class hierarchy. It just makes some methods/traits conditionally implemented based on the parameterized type. If anything, it keeps the class hierarchy as it was rather than adding wrapper types.

Reply:
Yeah, looks good. Will need to update the description in the module-level docs now that `InvoicePayer` no longer directly implements `EventHandler`.

Reply:
Sure thing, happy to. We could also do the other approach here: https://github.com/lightningdevkit/rust-lightning/pull/1787#discussion_r1013460196